include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +96,182 @@ parse(
-   corresponding function overload
-   `bool parse(char const*& it, char const* const end, error_code&, Arg_i const& t)`
-   should be defined to consume the input from [@ref it, @ref end) into `arg_i`.
+   For each type `Arg_i` in [`Arg_0`, `Arg_1`, ..., `Arg_n`] in `Args`, a
+   corresponding function overload of `parse`
+   `bool parse(char const*& it, char const* const end, error_code&, Arg_i const& arg_i)`
+   should be defined to consume the input from [`it`, `end`) into `arg_i`.<br>
 
-   This `parse` function is defined for `char`s so that it matches the BNF
-   string literals in [@ref it, @ref end).
-
-   The function returns `true` and resets @ref ec when the input is
+   The function returns `true` and resets `ec` when the input is
    parsed successfully. If the input cannot be parsed into the BNF elements,
-   the function returns `false` and sets the error number in @ref ec to the
+   the function returns `false` and sets the error number in `ec` to the
    appropriate error number defined in @ref boost::urls::error.
 
    @par Example
    @code
-   #include <iostream>
-   #include <string>
-
-   #include <boost/url/bnf/char_set.hpp>
-   #include <boost/url/bnf/parse.hpp>
-   #include <boost/url/error.hpp>
-
-   namespace boost {
-   namespace urls {
-   namespace detail {
-   // BNF element to parse a non-empty sequence of alpha chars ([a-zA-Z]+)
-   struct alpha_string_bnf {
-     // reference to a std::string where we should store the results
-     std::string &v;
-
-     // parsing an alpha_string_bnf element
-     friend bool parse(char const *&it, char const *end, error_code &ec,
-                       alpha_string_bnf &&t) {
-       // is_alpha functor
-       bnf::alpha_chars is_alpha;
-
-       // check if empty
-       if (it == end) {
-         ec = error::incomplete;
-         return false;
-       }
-
-       // check if not alpha at all
-       if (!is_alpha(*it)) {
-         ec = error::bad_alpha;
-         return false;
-       }
-       t.v.push_back(*it);
-       ++it;
-
-       // consume more alphas
-       while (it != end && is_alpha(*it)) {
-         t.v.push_back(*it);
-         ++it;
-       }
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (bnf::parse(it, end, ec, detail::alpha_string_bnf{alpha_l}, '.',
-                    detail::alpha_string_bnf{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
+   if (bnf::parse(it, end, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
```

Comment:
The correct usage is
```
using bnf::parse;
```

followed by calling `parse` without namespace qualification (like `swap`).

Reply:
> without namespace qualification

Reply:
I still see namespace qualification

Reply:
OK. I thought the customization point was only expected for the `char const*&, char const*, error_code&, T` overload, and the `bnf::parse` overloads were just entry points.

Reply:
it is easier to teach if we always write it uniformly